crates/bevy_animation/src/lib.rs

```diff
@@ -479,173 +539,99 @@ impl AnimationPlayer {
-    names: &Query<&Name>,
-    path_cache: &mut Vec<Option<Entity>>,
-) -> Option<Entity> {
-    // PERF: finding the target entity can be optimised
-    let mut current_entity = root;
-    path_cache.resize(path.parts.len(), None);
-
-    let mut parts = path.parts.iter().enumerate();
-
-    // check the first name is the root node which we already have
-    let Some((_, root_name)) = parts.next() else {
-        return None;
-    };
-    if names.get(current_entity) != Ok(root_name) {
-        return None;
-    }
-
-    for (idx, part) in parts {
-        let mut found = false;
-        let children = children.get(current_entity).ok()?;
-        if let Some(cached) = path_cache[idx] {
-            if children.contains(&cached) {
-                if let Ok(name) = names.get(cached) {
-                    if name == part {
-                        current_entity = cached;
-                        found = true;
-                    }
-                }
-            }
-        }
-        if !found {
-            for child in children.deref() {
-                if let Ok(name) = names.get(*child) {
-                    if name == part {
-                        // Found a children with the right name, continue to the next part
-                        current_entity = *child;
-                        path_cache[idx] = Some(*child);
-                        found = true;
-                        break;
-                    }
-                }
-            }
-        }
-        if !found {
-            warn!("Entity not found for path {:?} on part {:?}", path, part);
-            return None;
+/// A system that advances the time for all playing animations.
+pub fn advance_animations(
+    time: Res<Time>,
+    animation_clips: Res<Assets<AnimationClip>>,
+    mut players: Query<&mut AnimationPlayer>,
+) {
+    for mut player in players.iter_mut() {
+        let paused = player.paused;
+        if paused {
+            continue;
         }
-    }
-    Some(current_entity)
-}
 
-/// Verify that there are no ancestors of a given entity that have an [`AnimationPlayer`].
-fn verify_no_ancestor_player(
-    player_parent: Option<&Parent>,
-    parents: &Query<(Has<AnimationPlayer>, Option<&Parent>)>,
-) -> bool {
-    let Some(mut current) = player_parent.map(Parent::get) else {
-        return true;
-    };
-    loop {
-        let Ok((has_player, parent)) = parents.get(current) else {
-            return true;
+        // Advance the main animation.
+        if let Some(animation_clip) = animation_clips.get(&player.animation.animation_clip) {
+            player
+                .animation
+                .update(time.delta_seconds(), animation_clip.duration);
         };
-        if has_player {
-            return false;
-        }
-        if let Some(parent) = parent {
-            current = parent.get();
-        } else {
-            return true;
+
+        // Advance transition animations.
+        for transition in &mut player.transitions {
```

Comment:
What happens if advancing a transition makes the main animation change? Shouldn't we advance transitions first, then apply all animations which are still actually active (weight > 0)?

Reply:
I'm confused, how could advancing a transition affect the main animation?

Reply:
Maybe I'm misunderstanding what a transition is here?

Reply:
Wait no, we're updating `current_weight` below, so the question stands, what if it becomes 0?

Reply:
Although yes that won't change the main animation, if I understand what "main" means here

Reply:
I redid this code to expire the transitions eagerly once their weight hits 0.